### ISSUE TITLE
Inconsistency: limits on browser.alarms.create()

### ISSUE THREAD

**Issue by bershanskiy** [hober/web-extensions#422]
Updated: 31 May 2026.

## Background

`browser.alarms.create()` lets extensions register alarms, without any documented limits on their number and the extension-provided alarm names. All these names have to be stored somewhere, typically in a browser-internal database/registry. In theory, a buggy looped extension which dynamically generates alarm names can end up exceeding the limits of internal database. Less interestingly in practice, but more amusing, a humorous extension may implement an "unlimited" storage without having a `storage` or `unlimitedStorage` permission (by scheduling alarms far into the future so they don't go off and then retrieving their names).

## Current behavior

I tested the following:
 - creating many alarms in a loop
 - crating one alarms with a name equal to 10M with `chrome.alarms.create('a'.repeat(10_000_000), {when: 1_000_000_000_000_000});`

### Chromium
Chromium 117 ([PSA](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/8ZFwgCg-DvU/m/AxLjInQWAQAJ), [commit](https://github.com/chromium/chromium/commit/96b950168c0f85db3738e0215e5126fc2b305227)) limits each extension to 500 alarms, throwing an error `'An extension cannot have more than 500 active alarms.'` if extension attempts to create 501-th alarm.
Chromium does not limit the length of alarm names. In particular, I created an alarm with name equal to 100MB.

Chromium 144+ runs a histogram to determine this limit.

Chromium 148 logs a warning for alarm names longer than 1024 bytes announcing that Chrome 150 will throw on alarm names longer than 1024 bytes.

### Firefox
Firefox does not limit the number of alarms or lengths of alarm names created. ([Tracking bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1844605)) If too many alarms are created, then the offending extension hangs entirely (since event loop is swamped), the browser UI becomes unresponsive (or very badly responsive), and even if user closes the window process does not exit, pegging one CPU core at 100% utilization. User has to kill the process manually.
Firefox does not limit the length of alarm names. In particular, I created an alarm with name equal to 100MB.

## Desired behavior

I believe all browsers should have clear limits on the amount of data stored by alarms API, and ideally these limits would be identical across all browsers.

Proposed limits:
 - Extensions can store no more than 500 alarms at a time (500 is OK, 501 is an error). This behavior matches Chromium limit which was picked based on real-world test.
 - Every alarm name is no more than 1024 bytes long.
   - Chrome 148 and 149 log a warning for alarms over 1024 bytes long. ([commit](https://github.com/chromium/chromium/commit/432c30e3e1dcd03cecb869a1e23162d8ea1b9af2))
   - Chrome 150 will likely throw error for alarm with names longer than 1024 bytes. ([commit](https://github.com/chromium/chromium/commit/c0032deced7154bddd385a90cf094de5cb0b99a9))

**Comment by hanguokai** [comment]
Yes, Chrome [announced](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/8ZFwgCg-DvU/m/AxLjInQWAQAJ) the max-500 limit. I think it is Ok in practice.

Considering that the number has been limited, the name length is basically not a problem. Considering compatibility and less harm, it should be as large as possible, for example 1000.

**Comment by bershanskiy** [comment]
Hi, any sufficiently permissive length limit would be fine, as long as it actually exists. My main concern is that currently there is no limit at all and hence it is trivial to accidentally cause large writes to browser storage and hence overwhelm the internal DB. 500 keys each 1000 chars long would still be about 50KB of data which is manageable (really it becomes a problem only when keys get to GBs of data).

**Comment by xeenon** [comment]
Safari does not have a size or name length limit.

**Comment by zombie** [comment]
We're supportive of aligning with limits here, (500 alarms) as according to Google's stats it covers the 99.9% percentile.  Similarly, the max length of say 1000 characters also sounds reasonable.

**Comment by zombie** [comment]
I filed a Firefox bug to track this: https://bugzilla.mozilla.org/show_bug.cgi?id=1844605

**Comment by bershanskiy** [comment]
@hanguokai

> Considering compatibility and less harm, it should be as large as possible, for example 1000.

Is there any particular reason why extensions would want to store alarm names up to 1000 bytes long?

Google Chrome 148+ collects data about longest alarm names of each extension. [Data is available here.](https://github.com/w3c/webextensions/issues/935#issuecomment-3823905719) Since only 0.003% of extension installs have alarms longer than 1000 bytes, Chrome plans to restrict alarm lengths to 1024 bytes.

I would like to introduce even stricter limit and enforce it for alarms created with a new syntax described in #999. Based on Chrome's data, most extensions (> 99.8%) set alarm names no longer than 100 bytes. This limit of 100 bytes for each of 500 alarms sounds reasonable to me: decently light for mobile devices while being perfectly ergonomic for development. Bulk data should go to `storage.local` and sensitive data should go to `storage.session`. Would you oppose the limit of 100-byte alarm names?

**Comment by bershanskiy** [comment]
Chrome 150 introduced a hard limit for alarm names to be no more than 1024 bytes long. Chrome does not plan to further reduce this limit any time soon:[1]
>  I'm comfortable with saying 1024 is the final limit. This seems aggressive enough to curtail any significant abuse.

[1] https://chromium-review.googlesource.com/c/chromium/src/+/8270372?tab=comments
